sdk/core/azure-core/inc/azure/core/http/winhttp/win_http_client.hpp

```diff
@@ -6,26 +6,166 @@
  */
 
 #pragma once
+
 #ifdef BUILD_TRANSPORT_WINHTTP_ADAPTER
 
+#include "azure/core/context.hpp"
 #include "azure/core/http/http.hpp"
-#include "azure/core/http/policy.hpp"
+#include "azure/core/http/transport.hpp"
 
 #include <type_traits>
 #include <vector>
+#include <winhttp.h>
 
 namespace Azure { namespace Core { namespace Http {
 
+  namespace Details {
+
+    constexpr static int64_t DefaultUploadChunkSize = 1024 * 64;
+    constexpr static int64_t MaximumUploadChunkSize = 1024 * 1024;
+
+    struct HandleManager
+    {
+      Context const& m_context;
+      Request& m_request;
+      HINTERNET m_sessionHandle;
+      HINTERNET m_connectionHandle;
+      HINTERNET m_requestHandle;
+
+      HandleManager(Context const& context, Request& request)
+          : m_context(context), m_request(request)
+      {
+        m_sessionHandle = NULL;
+        m_connectionHandle = NULL;
+        m_requestHandle = NULL;
+      }
+
+      ~HandleManager()
+      {
+        // Close the handles and set them to null to avoid multiple calls to WinHTT to close the
+        // handles.
+        if (m_requestHandle)
+        {
+          WinHttpCloseHandle(m_requestHandle);
+          m_requestHandle = NULL;
+        }
+
+        if (m_connectionHandle)
+        {
+          WinHttpCloseHandle(m_connectionHandle);
+          m_connectionHandle = NULL;
+        }
+
+        if (m_sessionHandle)
+        {
+          WinHttpCloseHandle(m_sessionHandle);
+          m_sessionHandle = NULL;
+        }
+      }
+    };
+
+    class WinHttpStream : public BodyStream {
+    private:
+      std::unique_ptr<HandleManager> m_handleManager;
+      bool m_isEOF;
+
+      /**
+       * @brief This is a copy of the value of an HTTP response header `content-length`. The value
+       * is received as string and parsed to size_t. This field avoids parsing the string header
+       * every time from HTTP RawResponse.
+       *
+       * @remark This value is also used to avoid trying to read more data from network than what we
+       * are expecting to.
+       *
+       * @remark A value of -1 means the transfer encoding was chunked.
+       *
+       */
+      int64_t m_contentLength;
+
+      int64_t m_streamTotalRead;
+
+    public:
+      WinHttpStream(std::unique_ptr<HandleManager>& handleManager, int64_t contentLength)
+          : m_handleManager(std::move(handleManager)), m_contentLength(contentLength)
+      {
+        this->m_isEOF = false;
+        this->m_streamTotalRead = 0;
```

Comment:
What's the value of doing that? Assigning in the ctor conveys the intent more clearly to me. Do we generally set initial values of fields on types that have ctors in C++?

Reply:
Setting the members in the constructor body ends up as a waste if the value is the same as the default. For instance, setting false for Boolean. 
If you want to keep that clarity intention, set the members in the ctor declaration like:

WinHttpStream(std::unique_ptr<HandleManager>& handleManager, int64_t contentLength) : m_isEOF(false), m_streamTotalRead(0), ..., ..., ...

This way the member is init directly with the value that you want, instead of getting first a default and then the one you provide